waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +646,182 @@
+                 (concat new-instance-ids)
+                 (group-by #(if (contains? instance-id->heathly-instance %) :started :starting)))]
+        ;; Warn about any new instances for which we didn't track the scheduling time
+        (when (pos? unmatched-instance-count)
+          (log/warn unmatched-instance-count
+                    "untracked scheduled instances discovered in instance launch metric loop for service"
+                    service-id))
+        ;; Report schedule time for now-known instances
+        (doseq [start-time matched-start-times]
+          (let [duration (metrics/duration-between start-time current-time)]
+            (metrics/report-duration waiter-schedule-timer duration)
+            (metrics/report-duration service-schedule-timer duration)
+            #_(statsd/timer! service-schedule-timer duration)))
+        ;; Report startup time for instances that are current-time healthy
+        (doseq [instance-id started-instance-ids]
+          (let [{:keys [started-at]} (instance-id->heathly-instance instance-id)]
+            (metrics/report-duration service-startup-timer started-at current-time)))
+        ;; tracker-state'
+        (assoc tracker-state
+               :instance-counts instance-counts'
+               :instance-scheduling-start-times (vec instance-scheduling-start-times'')
+               :known-instance-ids known-instance-ids'
+               :starting-instance-ids (vec starting-instance-ids'))))))
+
+(defmacro build-intital-service-launch-trackers
+  [router-state-query-chan]
+  `(let [response-chan# (async/chan 1)
+         _# (async/>! ~router-state-query-chan response-chan#)
+         initial-router-state# (async/<! response-chan#)
+         iteration# (:iteration initial-router-state#)
+         initial-service-ids# (-> initial-router-state# :service-id->my-instance->slots keys set)
+         initial-service-id->launch-tracker#
+         (pc/for-map [service-id# initial-service-ids#]
+           service-id#
+           (let [instance-counts# (-> initial-router-state#
+                                      :service-id->instance-counts
+                                      (get service-id#))
+                 known-instance-ids# (->> initial-router-state#
+                                          ((juxt :service-id->healthy-instances
+                                                 :service-id->unhealthy-instances))
+                                          (map #(map :id (get % service-id#)))
+                                          (reduce into #{}))]
+             (make-launch-tracker service-id# instance-counts# known-instance-ids#)))]
+     (log/info "Starting launch metrics loop on iteration" iteration#
+               "with services:" initial-service-ids#)
+     {:known-service-ids initial-service-ids#
+      :previous-iteration iteration#
+      :service-id->launch-tracker initial-service-id->launch-tracker#}))
+
+(defn start-launch-metrics-maintainer
```

Comment:
The data flow is `scheduler/scheduler-syncer` -> `state/router-state-maintainer` -> `scheduler/launch-metrics-maintainer`. Can this data flow be `scheduler/scheduler-syncer` -> `scheduler/launch-metrics-maintainer` instead?

Reply:
@sradack you can be the tie-breaker here.

Reply:
We discussed offline, its possible that the maps do not need to be recomputed. @DaoWen is verifying the claim.